Mathlib/Algebra/Category/ModuleCat/Presheaf.lean

```diff
@@ -0,0 +1,154 @@
+/-
+Copyright (c) 2023 Scott Morrison All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import Mathlib.Algebra.Category.ModuleCat.Basic
+import Mathlib.Algebra.Category.Ring.Basic
+
+/-!
+# Presheaves of modules over a presheaf of rings.
+
+We give a hands-on description of a presheaf of modules over a fixed presheaf of rings,
+as a presheaf of abelian groups with additional data.
+
+## Future work
+
+* Show that we can check the sheaf condition at the level of abelian groups.
+* Compare this to the definition as a presheaf of pairs `(R, M)` with specified first part.
+* Compare this to the definition as a module object of the presheaf of rings
+  thought of as a monoid object.
+
+-/
+
+open CategoryTheory
+
+-- TODO: Porting note: move these definitions to the linear algebra library.
+-- They belong in `Mathlib.Algebra.Ring.CompTypeclasses`.
+
+class RingHomId {R : Type _} [Semiring R] (σ : R →+* R) : Prop where
+  eq_id : σ = RingHom.id R
+
+instance {R : Type _} [Semiring R] : RingHomId (RingHom.id R) where
+  eq_id := rfl
+
+/-- A generalisation of `LinearMap.id` that constructs the identity function
+as a `σ`-semilinear map for any ring homomorphism `σ` which we know is the identity. -/
+@[simps]
+def LinearMap.id' {R : Type _} [Semiring R]
+    {M : Type _} [AddCommMonoid M] [Module R M]
+    {σ : R →+* R} [RingHomId σ] : M →ₛₗ[σ] M where
+  toFun x := x
+  map_add' x y := rfl
+  map_smul' r x := by
+    have := (RingHomId.eq_id : σ = _)
+    subst this
+    rfl
+
+open LinearMap
+
+open Opposite
+
```

Comment:
Are we now in a post "declare universe v1 before u1" world or is this an oversight?